File: packages/core/tests/fixtures/typescript/comprehensive_definitions.ts
/**
 * Comprehensive TypeScript definitions test fixture
 * Tests all symbol definition types and edge cases
 */

// Basic functions
function simple_function() {
  return "hello";
}

const arrow_function = () => {
  return "arrow";
};

async function async_function(): Promise<string> {
  return "async";
}

function* generator_function(): Generator<number> {
  yield 1;
  yield 2;
}

// Generic functions
function generic_function<T>(param: T): T {
  return param;
}

function constrained_generic<T extends string>(param: T): T {
  return param;
}

// Variables and constants
const string_constant = "hello";
const number_variable = 42;
const hoisted_variable = "hoisted";

// Basic class
class SimpleClass {
  public name: string;
  private age: number;
  protected email: string;
  readonly id: string;

  constructor(name: string, age: number, email: string, id: string) {
    this.name = name;
    this.age = age;
    this.email = email;
    this.id = id;
  }

  public get_name(): string {
    return this.name;
  }

  private get_age(): number {
    return this.age;
  }

  protected get_email(): string {
    return this.email;
  }

  static create_default(): SimpleClass {
    return new SimpleClass("default", 0, "default@example.com", "default");
  }
}

// Generic class
class GenericClass<T, U> {
  private data: T;
  private metadata: U;

  constructor(data: T, metadata: U) {
    this.data = data;
    this.metadata = metadata;
  }

  get_data(): T {
    return this.data;
  }

  get_metadata(): U {
    return this.metadata;
  }

  static create<T, U>(data: T, metadata: U): GenericClass<T, U> {
    return new GenericClass(data, metadata);
  }
}

// Abstract class
abstract class AbstractBase {
  protected value: string = "";

  abstract process(): void;

  concrete(): string {
    return this.value;
  }
}

class ConcreteImplementation extends AbstractBase {
  process(): void {
    this.value = "processed";
  }
}

// Interfaces
interface BasicInterface {
  id: string;
  name: string;
}

interface GenericInterface<T> {
  data: T;
  process(input: T): T;
}

interface ExtendedInterface extends BasicInterface {
  description: string;
  get_value(): string;
}

// Interface with call signature
interface CallableInterface {
  (input: string): string;
  property: number;
}

// Interface with index signature
interface IndexableInterface {
  [key: string]: any;
  known_property: string;
}

// Type aliases
type StringAlias = string;
type GenericAlias<T> = T[];
type UnionType = string | number;
type IntersectionType = BasicInterface & { extra: boolean };
type ConditionalType<T> = T extends string ? true : false;
type MappedType<T> = {
  [K in keyof T]: T[K] | null;
};

// Enums
enum BasicEnum {
  FIRST,
  SECOND,
  THIRD
}

enum StringEnum {
  SUCCESS = "success",
  ERROR = "error",
  PENDING = "pending"
}

enum ComputedEnum {
  NONE = 0,
  READ = 1 << 1,
  WRITE = 1 << 2,
  READ_WRITE = READ | WRITE
}

const enum ConstEnum {
  A = "value_a",
  B = "value_b"
}

// Namespaces
namespace BasicNamespace {
  export interface NestedInterface {
    value: string;
  }

  export class NestedClass {
    constructor(public data: string) {}
  }

  export function nested_function(): string {
    return "nested";
  }

  export const nested_constant = "constant";

  export namespace InnerNamespace {
    export type InnerType = string;
    export const inner_value = 42;
  }
}

// Modules (TypeScript modules)
// declare module "external-module" {
//   export interface ExternalInterface {
//     method(): void;
//   }
//   export function externalFunction(): string;
// }

// Parameter properties
class ParameterProperties {
  constructor(
    public name: string,
    private age: number,
    protected email: string,
    readonly id: string,
  ) {}
}

// Method overloads
class OverloadedMethods {
  process(input: string): string;
  process(input: number): number;
  process(input: string | number): string | number {
    return input;
  }
}

// Decorators (if available)
function class_decorator(target: any) {
  return target;
}

function method_decorator(target: any, property_key: string, descriptor: PropertyDescriptor) {
  return descriptor;
}

function property_decorator(target: any, property_key: string) {
  // Property decorator
}

function parameter_decorator(target: any, property_key: string, parameter_index: number) {
  // Parameter decorator
}

@class_decorator
class DecoratedClass {
  @property_decorator
    decorated_property: string = "";

  @method_decorator
  decorated_method(@parameter_decorator param: string): string {
    return param;
  }
}

// Getters and setters
class GetterSetterClass {
  private _value: string = "";

  get value(): string {
    return this._value;
  }

  set value(new_value: string) {
    this._value = new_value;
  }
}

// Static blocks (ES2022)
class StaticBlockClass {
  static static_property: string;

  static {
    this.static_property = "initialized";
  }
}

// Complex nested structures
class OuterClass {
  private inner_value: string = "";

  public method(): void {
    // Local class
    class LocalClass {
      constructor(public data: string) {}

      process(): string {
        return this.data;
      }
    }

    // Local function
    function local_function(): string {
      // Local variable
      const local_var = "local";
      return local_var;
    }

    const instance = new LocalClass("test");
    const result = local_function();
  }
}

// Export statements
export { SimpleClass, GenericClass };
export type { BasicInterface, GenericInterface };
export { BasicEnum, StringEnum };
export default ConcreteImplementation;
export * from "./other-module";
export { SpecificExport } from "./specific-module";

File: packages/core/tests/fixtures/python/comprehensive_definitions.py
"""
Comprehensive Python definitions test fixture
Tests all symbol definition types and Python-specific behaviors
"""

# Module-level variables and constants
MODULE_CONSTANT = "constant"
module_variable = 42
_private_variable = "private"

# Basic functions
def simple_function():
    """Simple function definition"""
    return "hello"

def function_with_params(param1, param2="default"):
    """Function with parameters"""
    return param1 + param2

def function_with_annotations(param: str) -> str:
    """Function with type annotations"""
    return param.upper()

async def async_function() -> str:
    """Async function"""
    return "async"

def generator_function():
    """Generator function"""
    yield 1
    yield 2
    yield 3

# Lambda function
lambda_function = lambda x: x * 2

# Class definitions
class SimpleClass:
    """Simple class definition"""

    # Class variables
    class_variable = "class_var"
    _private_class_var = "private"

    def __init__(self, name, age):
        """Constructor method"""
        self.name = name  # Instance variable
        self.age = age
        self._private_instance = "private"

    def instance_method(self):
        """Instance method"""
        return self.name

    @classmethod
    def class_method(cls):
        """Class method"""
        return cls.class_variable

    @staticmethod
    def static_method():
        """Static method"""
        return "static"

    @property
    def age_property(self):
        """Property getter"""
        return self._age

    @age_property.setter
    def age_property(self, value):
        """Property setter"""
        self._age = value

# Inheritance
class BaseClass:
    """Base class for inheritance"""

    def __init__(self, value):
        self.value = value

    def base_method(self):
        return self.value

class DerivedClass(BaseClass):
    """Derived class"""

    def __init__(self, value, extra):
        super().__init__(value)
        self.extra = extra

    def base_method(self):
        """Override base method"""
        return f"derived: {self.value}"

    def derived_method(self):
        return self.extra

# Multiple inheritance
class Mixin:
    """Mixin class"""

    def mixin_method(self):
        return "mixin"

class MultipleInheritance(BaseClass, Mixin):
    """Class with multiple inheritance"""

    def __init__(self, value):
        super().__init__(value)

# Abstract base class (using abc module concepts)
class AbstractBase:
    """Abstract base class"""

    def concrete_method(self):
        return "concrete"

    def abstract_method(self):
        raise NotImplementedError("Must implement abstract_method")

# Dataclass-style class
class DataClassStyle:
    """Dataclass-style class"""

    def __init__(self, name: str, age: int, email: str = ""):
        self.name = name
        self.age = age
        self.email = email

# Nested classes
class OuterClass:
    """Outer class with nested definitions"""

    outer_var = "outer"

    class NestedClass:
        """Nested class"""

        nested_var = "nested"

        def __init__(self, data):
            self.data = data

        def nested_method(self):
            return self.data

    def outer_method(self):
        """Method with local definitions"""

        # Local variable
        local_var = "local"

        # Local function
        def local_function():
            return local_var + self.outer_var

        # Local class
        class LocalClass:
            def __init__(self, value):
                self.value = value

        return local_function()

# Decorators
def decorator_function(func):
    """Function decorator"""
    def wrapper(*args, **kwargs):
        result = func(*args, **kwargs)
        return f"decorated: {result}"
    return wrapper

def parametrized_decorator(param):
    """Parametrized decorator"""
    def decorator(func):
        def wrapper(*args, **kwargs):
            result = func(*args, **kwargs)
            return f"{param}: {result}"
        return wrapper
    return decorator

@decorator_function
def decorated_function():
    """Decorated function"""
    return "function"

@parametrized_decorator("custom")
def parametrized_decorated_function():
    """Function with parametrized decorator"""
    return "function"

# Class decorators
@decorator_function
class DecoratedClass:
    """Decorated class"""

    @decorator_function
    def decorated_method(self):
        return "method"

    @property
    @decorator_function
    def decorated_property(self):
        return "property"

# Exception classes
class CustomException(Exception):
    """Custom exception"""

    def __init__(self, message):
        super().__init__(message)
        self.message = message

class SpecificError(CustomException):
    """Specific error type"""
    pass

# Context managers
class ContextManager:
    """Context manager class"""

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        pass

# Magic methods
class MagicMethods:
    """Class with magic methods"""

    def __init__(self, value):
        self.value = value

    def __str__(self):
        return str(self.value)

    def __repr__(self):
        return f"MagicMethods({self.value})"

    def __add__(self, other):
        return MagicMethods(self.value + other.value)

    def __getitem__(self, key):
        return self.value[key]

    def __setitem__(self, key, value):
        self.value[key] = value

# Type variables and generics (typing module concepts)
from typing import TypeVar, Generic, List, Dict, Optional, Union

T = TypeVar('T')
K = TypeVar('K')
V = TypeVar('V')

class GenericClass(Generic[T]):
    """Generic class"""

    def __init__(self, data: T):
        self.data = data

    def get_data(self) -> T:
        return self.data

class GenericDict(Generic[K, V]):
    """Generic dictionary-like class"""

    def __init__(self):
        self._data: Dict[K, V] = {}

    def set_item(self, key: K, value: V) -> None:
        self._data[key] = value

    def get_item(self, key: K) -> Optional[V]:
        return self._data.get(key)

# Type aliases
StringList = List[str]
IntDict = Dict[str, int]
OptionalString = Optional[str]
UnionType = Union[str, int]

# Function with complex type annotations
def complex_typed_function(
    param1: List[str],
    param2: Dict[str, int],
    param3: Optional[str] = None,
    *args: str,
    **kwargs: int
) -> Union[str, None]:
    """Function with complex type annotations"""
    return param3

# Variables with type annotations
annotated_string: str = "hello"
annotated_list: List[int] = [1, 2, 3]
annotated_dict: Dict[str, str] = {"key": "value"}

# Global variables that test hoisting behavior
try:
    # This variable is defined conditionally
    conditional_variable = "conditional"
except:
    conditional_variable = "fallback"

# Variables in different scopes
if True:
    block_variable = "block"  # Not hoisted in Python

def scope_test_function():
    """Function to test variable scoping"""
    function_variable = "function"  # Local to function

    if True:
        nested_block_variable = "nested"  # Still function scope in Python

    return function_variable, nested_block_variable

# Import statements (would be at top in real code)
# These are here to test import handling
import os
import sys as system
from typing import Any, Callable
from collections import defaultdict, OrderedDict